app/Http/Controllers/DashboardController.php

```diff
@@ -67,185 +71,77 @@ public function dev_layout()
-            'site_id' => 'required|int|max:255',
-            'location_id' => 'required|int|max:255',
-            'device_id' => 'required|int|max:255',
-        ])->validate();
-        
+        $request->validate([
+            'site_id' => 'required|integer|digits_between:1,7',
+            'location_id' => 'sometimes|required|integer|digits_between:1,7',
+            'device_id' => 'sometimes|required|integer|digits_between:1,7',
+            'offset' => 'sometimes|required|integer|digits_between:1,7',
+        ]);
+    
         //Todo limit the amount of queries based on if there is changes
         //Check for changes in database since last update
-    
-        //Get the given site
-        $site = Site::select('id', 'name')->find($request->site_id);
-        //Get the given location
-        $location = Location::select('id', 'name', 'site_id')->find($request->location_id);
-        //Get the given device if it is still at the same location
-        $device = Device::publicDashData()->where('location_id', '=', $request->location_id)->find($request->device_id);
-        
-        if (empty($site))                 //Check if the site still exists
-        {
-            try
-            {
-                //Get the first site and if no sites exist throw an error
-                $site = Site::select('id', 'name')->firstOrFail();
-            }
-            catch(ModelNotFoundException $e)
-            {
-                //If no device is found then redirect the user to the home page and display a message about registering a device
-                return redirect('home')->with('no_devices', 'To access the dashboard page there must be at least one registered device assigned to a location.');
-            }
-            
-            //Get the first location from the site
-            $location = $site->locations()->select('id', 'name', 'site_id')->firstOrFail();
-            //Get the first device from the location
-            $device = $location->devices()->publicDashData()->firstOrFail();
-        }
-        else if (empty($location))        //Check if the location still exists
-        {
-            //Get the first location from the given site
-            $location = $site->locations()->select('id', 'name', 'site_id')->firstOrFail();
-            //Get the first device from the location
-            $device = $location->devices()->publicDashData()->firstOrFail();
-        }
-        else if (empty($device))        //Check if the device still exists at the given location
-        {
-            //Get the first device from the location
-            $device = $location->devices()->publicDashData()->firstOrFail();
-        }
         
-        $data = $this->dashData($site, $location, $device);
-        
-        return response()->json([ 'active_device' => $data[0], 'devices' => $data[1], 'locations' => $data[2], 'sites' => $data[3] ]);
-    }
+        //Limit the number of devices to be loaded
+        $limit = 4;
```

Comment:
This should be a setting somewhere or at the very least use the env() function to put this in the config.